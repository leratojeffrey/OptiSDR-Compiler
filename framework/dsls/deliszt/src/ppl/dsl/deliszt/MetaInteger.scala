package ppl.dsl.deliszt

/**
 * author: Michael Wu (mikemwu@stanford.edu)
 * last modified: 04/28/2011
 *
 * Pervasive Parallelism Laboratory (PPL)
 * Stanford University
 */

/*
 * MetaInteger.scala
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

abstract class IntM {
  type Suc <: IntM
  type Add[N <: IntM] <: IntM
}

object EnsureSize {
  implicit def EnsureSizeZero[T <: IntM] =
  new EnsureSize[Zero,Succ[T]] {
    def get(n : Int) : Int = n
  }

  implicit def EnsureSizeSucc[T <: IntM, TT <: IntM](implicit fn : EnsureSize[TT,T]) =
  new EnsureSize[Succ[TT],Succ[T]] {
    def get(n : Int) : Int = fn.get(n+1)
  }
}

//meta-function(?)(what is the term for this thing?)
//if IDX is less than SIZE, then an implicit object of type EnsureSize[IDX,SIZE] will
//exist with a idx function that returns the number IDX as an integer
abstract class EnsureSize[TT <: IntM,T <: IntM] {
  def get(n : Int) : Int
  def idx = get(0)
}

object MVal {
  implicit val MValZero = new MVal[Zero] {
    val value = 0
  }

  implicit def MValSucc[T <: IntM](implicit f: MVal[T]) = new MVal[Succ[T]] {
    val value = f.value + 1
  }
}

abstract class MVal[T] {
  val value: Int
}

class Zero extends IntM {
  type Suc = Succ[Zero]
  type Add[N <: IntM] = N
}

class Succ[N <: IntM] extends IntM {
  type Suc = Succ[Succ[N]]
  type Add[M <: IntM] = (N#Add[M])#Suc
}

object MetaInteger {
  type +[N1 <: IntM, N2 <: IntM] = N1#Add[N2]

  //Scala doesn't provide meta-integers like C++'s typename<int N>
  //so we have to define some integers here
  //this is a simplification of:
  //http://michid.wordpress.com/code/meta-programming-with-scala-part-i-addition/
  //somelike like Church encoding but the numbers are themselves objects and not
  //functions for simplicity since we don't need to define addition on them
  //as far I can tell ... if necessary I suppose we can define multiplcation
  //on these if we needed numbers higher than 9
  type _0 = Zero
  type _1 = Succ[_0]
  type _2 = Succ[_1]
  type _3 = Succ[_2]
  type _4 = Succ[_3]
  type _5 = Succ[_4]
  type _6 = Succ[_5]
  type _7 = Succ[_6]
  type _8 = Succ[_7]
  type _9 = Succ[_8]
  type _10 = Succ[_9]
  type _11 = Succ[_10]
  type _12 = Succ[_11]
  type _13 = Succ[_12]
  type _14 = Succ[_13]
  type _15 = Succ[_14]
  type _16 = Succ[_15]
  type _17 = Succ[_16]
  type _18 = Succ[_17]
  type _19 = Succ[_18]
  type _20 = Succ[_19]
  type _21 = Succ[_20]
  type _22 = Succ[_21]
  
  //for convience we provide objects representing each number as well
  //so you can avoid passing parameter arguments
  val _0 : _0 = new _0
  val _1 : _1 = new _1
  val _2 : _2 = new _2
  val _3 : _3 = new _3
  val _4 : _4 = new _4
  val _5 : _5 = new _5
  val _6 : _6 = new _6
  val _7 : _7 = new _7
  val _8 : _8 = new _8
  val _9 : _9 = new _9
  val _10 : _10 = new _10
  val _11 : _11 = new _11
  val _12 : _12 = new _12
  val _13 : _13 = new _13
  val _14 : _14 = new _14
  val _15 : _15 = new _15
  val _16 : _16 = new _16
  val _17 : _17 = new _17
  val _18 : _18 = new _18
  val _19 : _19 = new _19
  val _20 : _20 = new _20
  val _21 : _21 = new _21
  val _22 : _22 = new _22

  def MIntDepth[T <: IntM]()(implicit n: MVal[T]) = n.value
  def MIntDepth[T <: IntM](m : T)(implicit n: MVal[T]) = n.value
}
