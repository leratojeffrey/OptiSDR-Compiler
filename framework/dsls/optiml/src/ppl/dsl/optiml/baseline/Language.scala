package ppl.dsl.optiml.baseline

object LanguageDefs {
  val INITIAL_SEED = 100
  var randRef = new scala.util.Random(INITIAL_SEED)
  var _time = 0L
  
  def random(): Double = randRef.nextDouble
    
  def abs(d: Double) = implicitly[Numeric[Double]].abs(d) //java.lang.Math.abs(d)
  
  def sum[B:Numeric:Manifest](st: Int, en: Int)(f: Int => B) = {
    var acc = f(st)
    for (i <- st+1 until en) {
      acc = implicitly[Numeric[B]].plus(acc,f(i))
    }
    acc
  }
  
  def tic() {
    _time = System.currentTimeMillis()
  }
  
  def toc() {
    val now = System.currentTimeMillis()
    println("elapsed: " + (now-_time)/1000.0)
  }
}