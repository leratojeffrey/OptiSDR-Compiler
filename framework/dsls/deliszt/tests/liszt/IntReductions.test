import Liszt.Language._
import Liszt.MetaInteger._

@lisztcode
object IntReductions {
	val iField = FieldWithConst[Cell,Int](0)
	def main() {
		for(c <- cells(mesh)) {
			iField(c) += 2
		}
		for(c <- cells(mesh)) {
			iField(c) *= 2
		}
		for(c <- cells(mesh)) {
			iField(c) |= 1
		}
		for(c <- cells(mesh)) {
			iField(c) &= 7
		}
		for(c <- cells(mesh)) {
			iField(c) ^= 7
		}
		for(c <- cells(mesh)) {
			Print(iField(c))
		}
	}
}