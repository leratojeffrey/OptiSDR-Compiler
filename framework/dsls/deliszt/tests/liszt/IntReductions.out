5:c1-0 liszt: 2
5:c2-0 liszt: 2
5:c3-0 liszt: 2
5:c4-0 liszt: 2
5:c5-0 liszt: 2
5:c6-0 liszt: 2
5:c7-0 liszt: 2
5:c8-0 liszt: 2
5:c9-0 liszt: 2
5:c10-0 liszt: 2
5:c11-0 liszt: 2
5:c12-0 liszt: 2
5:c13-0 liszt: 2
5:c14-0 liszt: 2
5:c15-0 liszt: 2
5:c16-0 liszt: 2
5:c17-0 liszt: 2
5:c18-0 liszt: 2
5:c19-0 liszt: 2
5:c20-0 liszt: 2
5:c21-0 liszt: 2
5:c22-0 liszt: 2
5:c23-0 liszt: 2
5:c24-0 liszt: 2
5:c25-0 liszt: 2
5:c26-0 liszt: 2
5:c27-0 liszt: 2
5:c28-0 liszt: 2
5:c29-0 liszt: 2
5:c30-0 liszt: 2
5:c31-0 liszt: 2
5:c32-0 liszt: 2
5:c33-0 liszt: 2
5:c34-0 liszt: 2
5:c35-0 liszt: 2
5:c36-0 liszt: 2
5:c37-0 liszt: 2
5:c38-0 liszt: 2
5:c39-0 liszt: 2
5:c40-0 liszt: 2
5:c41-0 liszt: 2
5:c42-0 liszt: 2
5:c43-0 liszt: 2
5:c44-0 liszt: 2
5:c45-0 liszt: 2
5:c46-0 liszt: 2
5:c47-0 liszt: 2
5:c48-0 liszt: 2
5:c49-0 liszt: 2
5:c50-0 liszt: 2
5:c51-0 liszt: 2
5:c52-0 liszt: 2
5:c53-0 liszt: 2
5:c54-0 liszt: 2
5:c55-0 liszt: 2
5:c56-0 liszt: 2
5:c57-0 liszt: 2
5:c58-0 liszt: 2
5:c59-0 liszt: 2
5:c60-0 liszt: 2
5:c61-0 liszt: 2
5:c62-0 liszt: 2
5:c63-0 liszt: 2
5:c64-0 liszt: 2
5:c65-0 liszt: 2
5:c66-0 liszt: 2
5:c67-0 liszt: 2
5:c68-0 liszt: 2
5:c69-0 liszt: 2
5:c70-0 liszt: 2
5:c71-0 liszt: 2
5:c72-0 liszt: 2
5:c73-0 liszt: 2
5:c74-0 liszt: 2
5:c75-0 liszt: 2
5:c76-0 liszt: 2
5:c77-0 liszt: 2
5:c78-0 liszt: 2
5:c79-0 liszt: 2
5:c80-0 liszt: 2
5:c81-0 liszt: 2
5:c82-0 liszt: 2
5:c83-0 liszt: 2
5:c84-0 liszt: 2
5:c85-0 liszt: 2
5:c86-0 liszt: 2
5:c87-0 liszt: 2
5:c88-0 liszt: 2
5:c89-0 liszt: 2
5:c90-0 liszt: 2
5:c91-0 liszt: 2
5:c92-0 liszt: 2
5:c93-0 liszt: 2
5:c94-0 liszt: 2
5:c95-0 liszt: 2
5:c96-0 liszt: 2
5:c97-0 liszt: 2
5:c98-0 liszt: 2
5:c99-0 liszt: 2
5:c100-0 liszt: 2
5:c101-0 liszt: 2
5:c102-0 liszt: 2
5:c103-0 liszt: 2
5:c104-0 liszt: 2
5:c105-0 liszt: 2
5:c106-0 liszt: 2
5:c107-0 liszt: 2
5:c108-0 liszt: 2
5:c109-0 liszt: 2
5:c110-0 liszt: 2
5:c111-0 liszt: 2
5:c112-0 liszt: 2
5:c113-0 liszt: 2
5:c114-0 liszt: 2
5:c115-0 liszt: 2
5:c116-0 liszt: 2
5:c117-0 liszt: 2
5:c118-0 liszt: 2
5:c119-0 liszt: 2
5:c120-0 liszt: 2
5:c121-0 liszt: 2
5:c122-0 liszt: 2
5:c123-0 liszt: 2
5:c124-0 liszt: 2
5:c125-0 liszt: 2
5:c126-0 liszt: 2
5:c127-0 liszt: 2
5:c128-0 liszt: 2
5:c129-0 liszt: 2
5:c130-0 liszt: 2
5:c131-0 liszt: 2
5:c132-0 liszt: 2
5:c133-0 liszt: 2
5:c134-0 liszt: 2
5:c135-0 liszt: 2
5:c136-0 liszt: 2
5:c137-0 liszt: 2
5:c138-0 liszt: 2
5:c139-0 liszt: 2
5:c140-0 liszt: 2
5:c141-0 liszt: 2
5:c142-0 liszt: 2
5:c143-0 liszt: 2
5:c144-0 liszt: 2
5:c145-0 liszt: 2
5:c146-0 liszt: 2
5:c147-0 liszt: 2
5:c148-0 liszt: 2
5:c149-0 liszt: 2
5:c150-0 liszt: 2
5:c151-0 liszt: 2
5:c152-0 liszt: 2
5:c153-0 liszt: 2
5:c154-0 liszt: 2
5:c155-0 liszt: 2
5:c156-0 liszt: 2
5:c157-0 liszt: 2
5:c158-0 liszt: 2
5:c159-0 liszt: 2
5:c160-0 liszt: 2
5:c161-0 liszt: 2
5:c162-0 liszt: 2
5:c163-0 liszt: 2
5:c164-0 liszt: 2
5:c165-0 liszt: 2
5:c166-0 liszt: 2
5:c167-0 liszt: 2
5:c168-0 liszt: 2
5:c169-0 liszt: 2
5:c170-0 liszt: 2
5:c171-0 liszt: 2
5:c172-0 liszt: 2
5:c173-0 liszt: 2
5:c174-0 liszt: 2
5:c175-0 liszt: 2
5:c176-0 liszt: 2
5:c177-0 liszt: 2
5:c178-0 liszt: 2
5:c179-0 liszt: 2
5:c180-0 liszt: 2
5:c181-0 liszt: 2
5:c182-0 liszt: 2
5:c183-0 liszt: 2
5:c184-0 liszt: 2
5:c185-0 liszt: 2
5:c186-0 liszt: 2
5:c187-0 liszt: 2
5:c188-0 liszt: 2
5:c189-0 liszt: 2
5:c190-0 liszt: 2
5:c191-0 liszt: 2
5:c192-0 liszt: 2
5:c193-0 liszt: 2
5:c194-0 liszt: 2
5:c195-0 liszt: 2
5:c196-0 liszt: 2
5:c197-0 liszt: 2
5:c198-0 liszt: 2
5:c199-0 liszt: 2
5:c200-0 liszt: 2
5:c201-0 liszt: 2
5:c202-0 liszt: 2
5:c203-0 liszt: 2
5:c204-0 liszt: 2
5:c205-0 liszt: 2
5:c206-0 liszt: 2
5:c207-0 liszt: 2
5:c208-0 liszt: 2
5:c209-0 liszt: 2
5:c210-0 liszt: 2
5:c211-0 liszt: 2
5:c212-0 liszt: 2
5:c213-0 liszt: 2
5:c214-0 liszt: 2
5:c215-0 liszt: 2
5:c216-0 liszt: 2
5:c217-0 liszt: 2
5:c218-0 liszt: 2
5:c219-0 liszt: 2
5:c220-0 liszt: 2
5:c221-0 liszt: 2
5:c222-0 liszt: 2
5:c223-0 liszt: 2
5:c224-0 liszt: 2
5:c225-0 liszt: 2
5:c226-0 liszt: 2
5:c227-0 liszt: 2
5:c228-0 liszt: 2
5:c229-0 liszt: 2
5:c230-0 liszt: 2
5:c231-0 liszt: 2
5:c232-0 liszt: 2
5:c233-0 liszt: 2
5:c234-0 liszt: 2
5:c235-0 liszt: 2
5:c236-0 liszt: 2
5:c237-0 liszt: 2
5:c238-0 liszt: 2
5:c239-0 liszt: 2
5:c240-0 liszt: 2
5:c241-0 liszt: 2
5:c242-0 liszt: 2
5:c243-0 liszt: 2
5:c244-0 liszt: 2
5:c245-0 liszt: 2
5:c246-0 liszt: 2
5:c247-0 liszt: 2
5:c248-0 liszt: 2
5:c249-0 liszt: 2
5:c250-0 liszt: 2
